### ISSUE TITLE
Foreign method questions

### ISSUE THREAD

**Issue by kaadmy** [munificent/wren#408]
This is a question, not really an issue...

I've been struggling with passing compound (?) arguments to a foreign method in C++ from Wren, I can't find any documentation that shows how to pass anything other than lists and basic types.

Another thing that I can't figure out is accessing variables from the scope the foreign C++ function was called from in Wren.

Example of what I need to do:

```javascript
class Thing {
  construct new() {
    test_()

    System.print(_foo)
  }

  foreign test() // foreign function that sets _foo in this class instance
}

var inst = Thing.new()
```

**Comment by JonesAndrew** [comment]
My understanding is that to be able to set `_foo` in a foreign method you will need add a setter for `foo`, like `foo=(value)`, and call that from the c++ function. To get a access to a wren object in c++ you use the `WrenHandle* wrenGetSlotHandle(WrenVM* vm, int slot)` from `wren.h`.

**Comment by minirop** [comment]
With a foreign method, you can "create" a member. The easiest would be `_foo = test()`.

**Comment by kaadmy** [comment]
Can you have "nested" wrenInterprets?

For example call wrenInterpret from a method callback function to run Wren
code from the current scope?

**Comment by JonesAndrew** [comment]
I don't think it's possible to call a wrenInterpret in a certain scope, but as minirop mentioned you should not need to if you change the constructor to 
```dart
construct new() {
    _foo = test()
}
```

**Comment by kaadmy** [comment]
Okay, that'll work for my purpose, thanks!

**Comment by minirop** [comment]
it is possible (to some extends), that's how Meta.eval() works. https://github.com/munificent/wren/blob/master/src/optional/wren_opt_meta.c#L10

**Comment by munificent** [comment]
Yes, like others note, you can't directly access a field (read or write) from with a foreign function. There's two reasons for that:

1. I want to maintain the freedom to change how fields are stored in the VM. Generally, exposing something in the FFI ties your hands as a VM maintainer and prevents future optimizations. I'd like to avoid that when possible.

2. You'd probably want to look the field up by name. That would be slow and also require the VM to hold onto some data that it doesn't currently need to track. Today, fields are compiled directly to offset lookups in the instance data and at runtime, the field names are no longer around.

You can do like @JonesAndrew suggests and have the Wren code copy the foreign method's result into the field. If you do want the state stored and accessible from C, the other option is to make it a foreign *class*.
